File: chromium/chrome/browser/upboarding/query_tiles/android/tile_provider_bridge.cc
// Copyright 2020 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "chrome/browser/upboarding/query_tiles/android/tile_provider_bridge.h"

#include <memory>
#include <string>
#include <vector>

#include "base/android/callback_android.h"
#include "base/android/jni_string.h"
#include "chrome/browser/upboarding/query_tiles/jni_headers/TileProviderBridge_jni.h"
#include "ui/gfx/android/java_bitmap.h"
#include "ui/gfx/image/image.h"

using base::android::AttachCurrentThread;
using base::android::ConvertJavaStringToUTF8;
using base::android::ConvertUTF8ToJavaString;

namespace upboarding {

namespace {
const char kTileProviderBridgeKey[] = "tile_provider_bridge";

ScopedJavaLocalRef<jobject> createJavaTileAndMaybeAddToList(
    JNIEnv* env,
    ScopedJavaLocalRef<jobject> jlist,
    QueryTileEntry* tile) {
  ScopedJavaLocalRef<jobject> jchildren =
      Java_TileProviderBridge_createList(env);

  for (const auto& subtile : tile->sub_tiles)
    createJavaTileAndMaybeAddToList(env, jchildren, subtile.get());

  return Java_TileProviderBridge_createTileAndMaybeAddToList(
      env, jlist, ConvertUTF8ToJavaString(env, tile->id),
      ConvertUTF8ToJavaString(env, tile->display_text),
      ConvertUTF8ToJavaString(env, tile->accessibility_text),
      ConvertUTF8ToJavaString(env, tile->query_text), jchildren);
}

ScopedJavaLocalRef<jobject> createJavaTiles(
    JNIEnv* env,
    const std::vector<QueryTileEntry*>& tiles) {
  ScopedJavaLocalRef<jobject> jlist = Java_TileProviderBridge_createList(env);

  for (QueryTileEntry* tile : tiles)
    createJavaTileAndMaybeAddToList(env, jlist, tile);

  return jlist;
}

void RunGetTilesCallback(const JavaRef<jobject>& j_callback,
                         const std::vector<QueryTileEntry*>& tiles) {
  JNIEnv* env = AttachCurrentThread();
  RunObjectCallbackAndroid(j_callback, createJavaTiles(env, tiles));
}

void RunGeVisualsCallback(const JavaRef<jobject>& j_callback,
                          const gfx::Image& image) {
  ScopedJavaLocalRef<jobject> j_bitmap =
      gfx::ConvertToJavaBitmap(image.ToSkBitmap());
  RunObjectCallbackAndroid(j_callback, j_bitmap);
}

}  // namespace

// static
ScopedJavaLocalRef<jobject> TileProviderBridge::GetBridgeForTileService(
    TileService* tile_service) {
  if (!tile_service->GetUserData(kTileProviderBridgeKey)) {
    tile_service->SetUserData(
        kTileProviderBridgeKey,
        std::make_unique<TileProviderBridge>(tile_service));
  }

  TileProviderBridge* bridge = static_cast<TileProviderBridge*>(
      tile_service->GetUserData(kTileProviderBridgeKey));

  return ScopedJavaLocalRef<jobject>(bridge->java_obj_);
}

TileProviderBridge::TileProviderBridge(TileService* tile_service)
    : tile_service_(tile_service) {
  DCHECK(tile_service_);
  JNIEnv* env = base::android::AttachCurrentThread();
  java_obj_.Reset(
      env, Java_TileProviderBridge_create(env, reinterpret_cast<int64_t>(this))
               .obj());
}

TileProviderBridge::~TileProviderBridge() {
  JNIEnv* env = base::android::AttachCurrentThread();
  Java_TileProviderBridge_clearNativePtr(env, java_obj_);
}

void TileProviderBridge::GetQueryTiles(JNIEnv* env,
                                       const JavaParamRef<jobject>& jcaller,
                                       const JavaParamRef<jobject>& jcallback) {
  tile_service_->GetQueryTiles(base::BindOnce(
      &RunGetTilesCallback, ScopedJavaGlobalRef<jobject>(jcallback)));
}

void TileProviderBridge::GetVisuals(JNIEnv* env,
                                    const JavaParamRef<jobject>& jcaller,
                                    const JavaParamRef<jstring>& jid,
                                    const JavaParamRef<jobject>& jcallback) {
  std::string tile_id = ConvertJavaStringToUTF8(env, jid);
  tile_service_->GetVisuals(
      tile_id, base::BindOnce(&RunGeVisualsCallback,
                              ScopedJavaGlobalRef<jobject>(jcallback)));
}

}  // namespace upboarding

File: chromium/chrome/browser/upboarding/query_tiles/android/java/src/org/chromium/chrome/browser/query_tiles/bridges/TileProviderBridge.java
// Copyright 2020 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.chrome.browser.query_tiles.bridges;

import android.graphics.Bitmap;

import androidx.annotation.Nullable;

import org.chromium.base.Callback;
import org.chromium.base.annotations.CalledByNative;
import org.chromium.base.annotations.JNINamespace;
import org.chromium.base.annotations.NativeMethods;
import org.chromium.chrome.browser.query_tiles.Tile;
import org.chromium.chrome.browser.query_tiles.TileProvider;

import java.util.ArrayList;
import java.util.List;

/**
 * Bridge to the native query tile service for the given {@link Profile}.
 */
@JNINamespace("upboarding")
public class TileProviderBridge implements TileProvider {
    private long mNativeTileProviderBridge;

    @CalledByNative
    private static TileProviderBridge create(long nativePtr) {
        return new TileProviderBridge(nativePtr);
    }

    private TileProviderBridge(long nativePtr) {
        mNativeTileProviderBridge = nativePtr;
    }

    @CalledByNative
    private void clearNativePtr() {
        mNativeTileProviderBridge = 0;
    }

    @Override
    public void getQueryTiles(Callback<List<Tile>> callback) {
        if (mNativeTileProviderBridge == 0) return;
        TileProviderBridgeJni.get().getQueryTiles(mNativeTileProviderBridge, this, callback);
    }

    @Override
    public void getVisuals(String id, Callback<List<Bitmap>> callback) {
        if (mNativeTileProviderBridge == 0) return;
        TileProviderBridgeJni.get().getVisuals(mNativeTileProviderBridge, this, id, callback);
    }

    @CalledByNative
    private static List<Tile> createList() {
        return new ArrayList<>();
    }

    @CalledByNative
    private static Tile createTileAndMaybeAddToList(@Nullable List<Tile> list, String tileId,
            String displayTitle, String accessibilityText, String queryText, List<Tile> children) {
        Tile tile = new Tile(tileId, displayTitle, accessibilityText, queryText, children);
        if (list != null) list.add(tile);
        return tile;
    }

    @NativeMethods
    interface Natives {
        void getQueryTiles(long nativeTileProviderBridge, TileProviderBridge caller,
                Callback<List<Tile>> callback);
        void getVisuals(long nativeTileProviderBridge, TileProviderBridge caller, String id,
                Callback<List<Bitmap>> callback);
    }
}